parser.lisp

```diff
@@ -20,26 +20,111 @@
+               (digit-char-p (elt string 1))))
+      (with-output-to-string (s)
+        (flet ((escape-regular-string (string)
+                 (dotimes (index (length string))
+                   (let ((char (elt string index)))
+                     (cond
+                       ((char= #\Nul char)
+                        (write-string "\\fffd " s))
+                       ((or (<= #x1 (char-code char) #x1f)
+                            (= (char-code char) #x7f))
+                        (format s "\\~X " (char-code char)))
+                       ((escapable char)
+                        (write-char #\\ s)
+                        (write-char char s))
+                       (t
+                        (write-char char s)))))))
+          ;; Process first char.
+          (cond
+            ((and (= (length string) 1)
+                  (char= #\- (elt string 0)))
+             (write-char #\\ s)
+             (write-char (elt string 0) s))
+            ((and (= (length string) 1)
+                  (digit-char-p (elt string 0)))
+             (format s "\\~X " (char-code (elt string 0))))
+            (t
+             (escape-regular-string (subseq string 0 1))))
+          ;; Second char.
+          (when (>= (length string) 2)
+            (if (and (char= #\- (elt string 0))
+                     (digit-char-p (elt string 1)))
+                (format s "\\~X " (char-code (elt string 1)))
+                (escape-regular-string (subseq string 1 2))))
+          ;; All the rest.
+          (when (> (length string) 2)
+            (escape-regular-string (subseq string 2)))))
+      string))
+
+(defun css-unescape (string)
+  "Get the original contents of the escaped STRING."
+  (declare (optimize speed)
+           (type string string))
+  (if (search "\\" string)
+      (with-output-to-string (s)
+        (do* ((index 0 (1+ index))
+              (char (elt string index) (elt string index))
+              (next-char (when (< index (1- (length string)))
+                           (elt string (1+ index)))
+                         (when (< index (1- (length string)))
+                           (elt string (1+ index)))))
+             (nil)
+          (cond
+            ((search "\\fffd " string
+                     :start2 index :end2 (min (length string)
+                                              (+ index 6)))
+             (write-char #\Nul s)
+             (setf index (position #\Space string :start (1+ index))))
+            ((and (eql #\\ char)
+                  (digit-char-p next-char 16))
+             (write-char (code-char (parse-integer string :radix 16 :start (1+ index))) s)
+             (setf index (position #\Space string :start (1+ index))))
+            ((eql #\\ char)
+             (write-char next-char s)
+             (incf index))
+            (t
+             (write-char char s)))
+          (when (>= index (1- (length string)))
+            (return-from css-unescape (get-output-stream-string s)))))
```

Comment:
Just `(return)` should be enough. It'll escape the `do*` and `with-output-to-string` will return the buffer.